docs/best-practices.md

```diff
@@ -0,0 +1,799 @@
+
+```rust
+#[near_bindgen]
+impl Contract {
+    #[private]
+    pub fn resolve_transfer(&mut self) {
+        env::log_str("This is a callback");
+    }
+}
+```
+
+This is equivalent to:
+
+```rust
+#[near_bindgen]
+impl Contract {
+    pub fn resolve_transfer(&mut self) {
+        if env::current_account_id() != env::predecessor_account_id() {
+            near_sdk::env::panic(b"Method resolve_transfer is private");
+        }
+        env::log_str("This is a callback");
+    }
+}
+```
+
+## Integer JSON types
+
+NEAR Protocol currently expects contracts to support JSON serialization. JSON can't handle large integers (above `2**53` bits).
+That's why, for all exported methods, you should use helper classes from the `json_types` in `near_sdk` instead of `u64` and `u128`.
+We provide types `U64` and `U128`, which wrap the integer into a struct and implement JSON serialization and
+deserialization as a base-10 strings.
+
+You can convert from `U64` to `u64` and back using `std::convert::Into`, e.g.
+
+```rust
+#[near_bindgen]
+impl Contract {
+    pub fn mult(&self, a: U64, b: U64) -> U128 {
+        let a: u64 = a.into();
+        let b: u64 = b.into();
+        let product = u128::from(a) * u128::from(b);
+        product.into()
+    }
+}
+```
+
+You can also access inner values and using `.0`:
+
+```diff
+ #[near_bindgen]
+ impl Contract {
+     pub fn mult(&self, a: U64, b: U64) -> U128 {
+-        let a: u64 = a.into();
++        let a = a.0;
+-        let b: u64 = b.into();
++        let b = b.0;
+         let product = u128::from(a) * u128::from(b);
+         product.into()
+     }
+ }
+```
+
+And you can cast the lower-case `u` variants to upper-case `U` variants using `U64(...)` and `U128(...)`:
+
+```diff
+ #[near_bindgen]
+ impl Contract {
+     pub fn mult(&self, a: U64, b: U64) -> U128 {
+         let a = a.0;
+         let b = b.0;
+         let product = u128::from(a) * u128::from(b);
+-        product.into()
++        U128(product)
+     }
+ }
+```
+
+Combining it all:
+
+```rust
+#[near_bindgen]
+impl Contract {
+    pub fn mult(&self, a: U64, b: U64) -> U128 {
+        U128(u128::from(a.0) * u128::from(b.0))
+    }
+}
+```
+
+## `Base64VecU8` JSON type
+
+Contracts often need to receive or return binary data.
+Encoding a `Vec<u8>` with JSON will lead to an integer array, e.g. `[110, 101, 97, 114]`
+This is inefficient in both compute and space.
+
+`Base64VecU8` is a wrapper on top of `Vec<u8>` that allows to pass it as arguments or result.
+
+```rust
+#[near_bindgen]
+#[derive(BorshDeserialize, BorshSerialize, PanicOnDefault)]
+pub struct Contract {
+    // Notice, internally we store `Vec<u8>` 
+    pub data: Vec<u8>,
+}
+
+#[near_bindgen]
+impl Contract {
+    #[init]
+    pub fn new(data: Base64VecU8) -> Self {
+        Self {
+            data: data.into(),
+        }
+    }
+
+    pub fn get_data(self) -> Base64VecU8 {
+        self.data.into()
+    }
+}
+```
```

Comment:
All covered in https://sdk-docs-pr-37.onrender.com/contract-structure/near-bindgen#initialization-methods

Unless you want to add any of these details to it

Reply:
Hm, I'm not seeing this info covered in the SDK docs yet. It feels like we can keep this section to me

Reply:
Whoops, was searching only Rust files, will look again…

Reply:
migrated: https://github.com/near/sdk-docs/pull/37/commits/88022f03f4e4522f1a427d3d78135693dccfb7eb